controllers/openstacknet_controller.go

```diff
@@ -93,69 +98,287 @@ func (r *OpenStackNetReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error
 		return ctrl.Result{}, err
 	}
 
-	// generate NodeNetworkConfigurationPolicy
-	ncp := common.NetworkConfigurationPolicy{
-		Name:                           instance.Name,
-		Labels:                         common.GetLabelSelector(instance, openstacknet.AppLabel),
-		NodeNetworkConfigurationPolicy: instance.Spec.AttachConfiguration.NodeNetworkConfigurationPolicy,
+	// examine DeletionTimestamp to determine if object is under deletion
+	if instance.ObjectMeta.DeletionTimestamp.IsZero() {
+		// The object is not being deleted, so if it does not have our finalizer,
+		// then lets add the finalizer and update the object. This is equivalent
+		// registering our finalizer.
+		if !controllerutil.ContainsFinalizer(instance, openstacknet.FinalizerName) {
+			controllerutil.AddFinalizer(instance, openstacknet.FinalizerName)
+			if err := r.Update(context.Background(), instance); err != nil {
+				return ctrl.Result{}, err
+			}
+			r.Log.Info(fmt.Sprintf("Finalizer %s added to CR %s", openstacknet.FinalizerName, instance.Name))
+		}
+	} else {
+		// 1. check if finalizer is there
+		// Reconcile if finalizer got already removed
+		if !controllerutil.ContainsFinalizer(instance, openstacknet.FinalizerName) {
+			return ctrl.Result{}, nil
+		}
+
+		// 2. Clean up resources used by the operator
+		// SRIOV resources
+		err = r.sriovResourceCleanup(instance)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		// 3. as last step remove the finalizer on the operator CR to finish delete
+		controllerutil.RemoveFinalizer(instance, openstacknet.FinalizerName)
+		err = r.Client.Update(context.TODO(), instance)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		r.Log.Info(fmt.Sprintf("CR %s deleted", instance.Name))
+		return ctrl.Result{}, nil
 	}
-	err = common.CreateOrUpdateNetworkConfigurationPolicy(r, instance, instance.Kind, &ncp)
-	if err != nil {
-		return ctrl.Result{}, err
+
+	// If we have an SRIOV definition on this network, use that.  Otherwise assume the
+	// non-SRIOV configuration must be present
+	if instance.Spec.AttachConfiguration.NodeSriovConfigurationPolicy.DesiredState.Port != "" {
+		if err := r.ensureSriov(instance); err != nil {
+			return ctrl.Result{}, err
+		}
+	} else {
+		// generate NodeNetworkConfigurationPolicy
```

Comment:
like with ensureSriov, we might want to move all the bellow to a method in a follow up patch